c2rust-analyze/src/main.rs

```diff
@@ -35,29 +41,113 @@ mod pointer_id;
+        if self.0.is_some() {
+            panic!("value is already set");
+        }
+        self.0 = Some(x);
+    }
+
+    pub fn clear(&mut self) {
+        if self.0.is_none() {
+            panic!("value is already cleared");
+        }
+        self.0 = None;
+    }
+
+    pub fn get(&self) -> &T {
+        self.0.as_ref().expect("value is not set")
+    }
+
+    pub fn get_mut(&mut self) -> &mut T {
+        self.0.as_mut().expect("value is not set")
+    }
+
+    pub fn take(&mut self) -> T {
+        self.0.take().expect("value is not set")
+    }
+}
+
+impl<T> Deref for MaybeUnset<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        self.get()
+    }
+}
+
+impl<T> DerefMut for MaybeUnset<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.get_mut()
+    }
+}
+
 fn run(tcx: TyCtxt) {
     let mut gacx = GlobalAnalysisCtxt::new(tcx);
-    let mut func_info = Vec::new();
+    let mut func_info = HashMap::new();
+
+    #[derive(Default)]
+    struct FuncInfo<'tcx> {
+        acx_data: MaybeUnset<AnalysisCtxtData<'tcx>>,
+        dataflow: MaybeUnset<DataflowConstraints>,
+        l_equiv: MaybeUnset<LocalEquivSet>,
+        lasn: MaybeUnset<LocalAssignment>,
+    }
 
     // Initial pass to gather equivalence constraints, which state that two pointer types must be
     // converted to the same reference type.  Some additional data computed during this the process
     // is kept around for use in later passes.
 
     // TODO: assign global PointerIds
+    for ldid in tcx.hir().body_owners() {
+        let sig = tcx.fn_sig(ldid.to_def_id());
+        let sig = tcx.erase_late_bound_regions(sig);
+
+        let inputs = sig
+            .inputs()
+            .iter()
+            .map(|&ty| gacx.assign_pointer_ids(ty))
+            .collect::<Vec<_>>();
+        let inputs = gacx.lcx.mk_slice(&inputs);
+        let output = gacx.assign_pointer_ids(sig.output());
+
+        let lsig = LFnSig { inputs, output };
+        gacx.fn_sigs.insert(ldid.to_def_id(), lsig);
+    }
 
     let mut g_equiv = GlobalEquivSet::new(gacx.num_pointers());
     for ldid in tcx.hir().body_owners() {
         let ldid_const = WithOptConstParam::unknown(ldid);
         let mir = tcx.mir_built(ldid_const);
         let mir = mir.borrow();
+        let lsig = gacx.fn_sigs.get(&ldid.to_def_id()).unwrap().clone();
 
         let mut acx = gacx.function_context(&mir);
 
         // Assign PointerIds to local types
         assert!(acx.local_tys.is_empty());
         acx.local_tys = IndexVec::with_capacity(mir.local_decls.len());
         for (local, decl) in mir.local_decls.iter_enumerated() {
-            let lty = assign_pointer_ids(&mut acx, decl.ty);
+            let lty = match mir.local_kind(local) {
+                LocalKind::Var | LocalKind::Temp => acx.assign_pointer_ids(decl.ty),
+                LocalKind::Arg => {
+                    debug_assert!(local.as_usize() >= 1 && local.as_usize() <= mir.arg_count);
```

Comment:
Why a `debug_assert!` and not just an `assert!`?  If the assert fails but it's in release mode, this will cause an overflow, right?

Reply:
If this condition doesn't hold, the next line will panic with index out of bounds (possibly after underflowing to `usize::MAX`).  The assertion here is technically redundant - its only purpose is to express the expected property in a more straightforward way.